File: Contests/ARC/072/fsol.py
n, l = list(map(int, input().strip().split()))

dq = []
ans = []
totx = 0
toty = 0

frt = 0
for i in range(n):
    # print(i, dq, frt)
    # print(frt)
    t, v = list(map(int, input().strip().split()))
    dq.append([v, t*v])
    # print(i, dq, frt)
    totx += v
    toty += t*v

    if i == 0:
        ans.append(dq[0][1])
        continue

    while True:
        x1 = dq[frt][0]
        y1 = dq[frt][1]
        # print(x1, y1, totx, toty)

        if totx - x1 < l:
            dq[frt][1] = y1 / x1 * (x1 - totx + l)
            dq[frt][0] = x1 - totx + l
            totx = totx - x1 + dq[frt][0]
            toty = toty - y1 + dq[frt][1]   
            ans.append(toty)
            break
        else:
            frt += 1
            # if frt > 0:
            #     print(i, frt)
            totx -= x1
            toty -= y1

    # print(i, dq, frt)

    x, y = dq.pop()
    while True:
        if (len(dq) == frt):
            dq.append([x, y])
            break
        x1 = dq[-1][0]
        y1 = dq[-1][1]
        if y1 / x1 < y / x:
            dq.append([x, y])
            break
        else:
            x += x1
            y += y1
            dq.pop()

for i in range(n):
    ans[i] = ans[i] / l
print('\n'.join(map(str, ans)))

File: Contests/ARC/072/fsol.cpp
#include <bits/stdc++.h>
using namespace std;
using namespace std::chrono;

typedef long long ll;
typedef vector<int> vi;
typedef pair<int, int> pii;
typedef vector<pii> vii;
typedef vector<vector<int> > vvi;
typedef vector<string> vs;
typedef complex<double> cd;
#define rep(i,l,r) for(int i=l;i<=r;++i)
#define per(i,r,l) for(int i=r;i>=l;--i)
#define rep0(i,l,r) for(int i=l;i<r;++i)
#define forn(i,n) for(int i=0;i<n;++i)
#define all(a) (a).begin(), (a).end()
#define allr(a) (a).rbegin(), (a).rend()
#define foreach(a) for(auto it = a.begin(); it != a.end(); it++)
#define mem(a,b) memset(a, (b), sizeof(a))

const int maxn = 1e5 + 5;
const ll MOD = 1e9 + 7; // 998244353
const ll INF = 1e9;
const int LOG = 26;
const char min_char = 'a';
const double EPS = 1e-9;
const double PI = 3.14159265358979323846;

bool compare(ll x, ll y, ll z, ll t){
    if (y / x > t / z) return false;
    if (y / x < t / z) return true;
    ll u = y % x, v = t % z;
    if (u * z <= v * x) return true;
    else return false;
}

void solve(){
    int n; cin >> n;
    ll l; cin >> l;
    deque<pair<ll, ll>> q;
    vector<ll> ans;
    ll totx = 0, toty = 0;
    for (int i = 0; i < n; i++){
        ll t, v;
        cin >> t >> v;
        if (i == 0){
            ans.push_back(t * v);
            q.push_front({v, t*v});
            totx += v; toty += t*v;
            continue;
        }
        else{
            q.push_front({v, t*v});
            totx += v;
            toty += t*v;

            // find cut V
            while (true){
                pair<ll, ll> cur = q.back();
                ll x1 = cur.first, y1 = cur.second;
                if (totx - x1 < l){
                    ans.push_back(y1);
                }
            }
            
            ll x = v, y = t*v;

            while (true){
                pair<ll, ll> cur = q.front();
                ll x1 = cur.first, y1 = cur.second;
                bool b = compare(x, y, x1, y1);
                if (b){
                    q.push_front({x, y});
                    break;
                }
                else{

                }
            }
        }
    }
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);
    auto start = high_resolution_clock::now();
    int T = 1;
    // cin >> T;
    while(T--){ 
        solve();
    }
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(stop - start);
    cerr << "time: " << duration.count() << "ms\n";
}